File: Fractals/Fractals/SunflowerComponent.vb
Imports System.Collections.Generic

Imports Grasshopper.Kernel
Imports Rhino.Geometry

'' In order to load the result of this wizard, you will also need to
'' add the output bin/ folder of this project to the list of loaded
'' folder in Grasshopper.
'' You can use the _GrasshopperDeveloperSettings Rhino command for that.

Public Class SunflowerComponent
    Inherits GH_Component
    ''' <summary>
    ''' Each implementation of GH_Component must provide a public 
    ''' constructor without any arguments.
    ''' Category represents the Tab in which the component will appear, 
    ''' Subcategory the panel. If you use non-existing tab or panel names, 
    ''' new tabs/panels will automatically be created.
    ''' </summary>
    Public Sub New()
        MyBase.New("Vogel Sunflower", "Vogel",
                "Vogel Sunflower Fractal",
                "Fractals", "Fractals")
    End Sub

    ''' <summary>
    ''' Registers all the input parameters for this component.
    ''' </summary>
    Protected Overrides Sub RegisterInputParams(pManager As GH_Component.GH_InputParamManager)
        pManager.AddIntegerParameter("Number of Iterations", "N", "Number of Iterations - Default value 3000", GH_ParamAccess.item, 3000)
        pManager.AddPointParameter("Generation point", "S", "Generation Point", GH_ParamAccess.item)
        pManager.AddIntegerParameter("Degree", "D", "Degree - Default value 5", GH_ParamAccess.item, 5)
    End Sub

    ''' <summary>
    ''' Registers all the output parameters for this component.
    ''' </summary>
    Protected Overrides Sub RegisterOutputParams(pManager As GH_Component.GH_OutputParamManager)
        pManager.AddPointParameter("Points", "P", "Points", GH_ParamAccess.list)
        pManager.AddNumberParameter("Radius", "R", "Radius of circles at points", GH_ParamAccess.list)
    End Sub

    ''' <summary>
    ''' This is the method that actually does the work.
    ''' </summary>
    ''' <param name="DA">The DA object can be used to retrieve data from input parameters and 
    ''' to store data in output parameters.</param>
    Protected Overrides Sub SolveInstance(DA As IGH_DataAccess)
        Dim Angle As Double = 0
        Dim R As Double = 0
        Dim c As Double = 0
        Dim pc As New List(Of Point)
        Dim radii As New List(Of Double)
        Dim x As Double = 0
        Dim y As Double = 0
        Dim numberofseeds As Integer = 3000
        Dim initial As New Point3d(0, 0, 0)
        Dim degree As Int32 = 5
        Dim trans As New Vector3d
        If (Not DA.GetData(0, numberofseeds)) Then Return
        If (Not DA.GetData(1, initial)) Then Return
        If (Not DA.GetData(2, degree)) Then Return
        If (degree <= 1) Then
            AddRuntimeMessage(GH_RuntimeMessageLevel.Error, "Degree must be greater than 1.")
            Return
        End If
        If (((Math.Sqrt(degree)) - (Math.Floor(Math.Sqrt(degree)))) = 0) Then
            AddRuntimeMessage(GH_RuntimeMessageLevel.Error, "Degree must not be a perfect square.")
            Return
        End If
        c = (Math.Sqrt(degree) + 1) / 2
        For i As Integer = 0 To numberofseeds
            R = i ^ c / (numberofseeds / 2)
            Angle = 2 * Math.PI * c * i
            x = initial.X + (R * Math.Sin(Angle) + (numberofseeds / 10))
            y = initial.Y + (R * Math.Cos(Angle) + (numberofseeds / 10))
            Dim p As New Point(New Point3d(x, y, initial.Z))
            If i = 0 Then
                Dim l As New Line(initial, p.Location)
                trans = ((l.UnitTangent * l.Length) * (-1))
            End If
            p.Translate(trans)
            pc.Add(p)
            Dim rad As Double = 0
            rad = i / (numberofseeds / 10)
            radii.Add(rad)
        Next
        DA.SetDataList("Radius", radii)
        DA.SetDataList("Points", pc)
    End Sub

    ''' <summary>
    ''' Provides an Icon for every component that will be visible in the User Interface.
    ''' Icons need to be 24x24 pixels.
    ''' </summary>
    Protected Overrides ReadOnly Property Icon() As System.Drawing.Bitmap
        Get
            'You can add image files to your project resources and access them like this:
            ' return Resources.IconForThisComponent;
            Return My.Resources.Sunflower
        End Get
    End Property

    ''' <summary>
    ''' Each component must have a unique Guid to identify it. 
    ''' It is vital this Guid doesn't change otherwise old ghx files 
    ''' that use the old ID will partially fail during loading.
    ''' </summary>
    Public Overrides ReadOnly Property ComponentGuid() As Guid
        Get
            Return New Guid("11aaa803-ae68-43c6-b516-50dbdebb2a79")
        End Get
    End Property
End Class

File: Fractals/SunflowerComponent.cs
using System;
using System.Collections.Generic;

using Grasshopper.Kernel;
using Rhino.Geometry;

namespace Fractals
{
    public class SunflowerComponent : GH_Component
    {
        /// <summary>
        /// Initializes a new instance of the SunflowerComponent class.
        /// </summary>
        public SunflowerComponent()
          : base("Vogel Sunflower", "Sunflower",
              "Vogel Sunflower Fractal",
              "Fractals", "2D")
        {
        }

        /// <summary>
        /// Registers all the input parameters for this component.
        /// </summary>
        protected override void RegisterInputParams(GH_Component.GH_InputParamManager pManager)
        {
            pManager.AddPlaneParameter("Plane", "P", "Plane", GH_ParamAccess.item, Plane.WorldXY);
            pManager.AddIntegerParameter("Number of Iterations", "N", "Number of Iterations", GH_ParamAccess.item, 3000);
            pManager.AddIntegerParameter("Degree", "D", "Degree", GH_ParamAccess.item, 5);
        }

        /// <summary>
        /// Registers all the output parameters for this component.
        /// </summary>
        protected override void RegisterOutputParams(GH_Component.GH_OutputParamManager pManager)
        {
            pManager.AddPlaneParameter("Points", "P", "Points", GH_ParamAccess.list);
            pManager.AddNumberParameter("Radius Multiplier", "R", "Radius Multiplier for circles at points", GH_ParamAccess.list);
        }

        /// <summary>
        /// This is the method that actually does the work.
        /// </summary>
        /// <param name="DA">The DA object is used to retrieve from inputs and store in outputs.</param>
        protected override void SolveInstance(IGH_DataAccess DA)
        {
            Plane plane = Plane.WorldXY;
            int iterations = 3000;
            int deg = 5;
            if (!DA.GetData(0, ref plane)) return;
            if (!DA.GetData(1, ref iterations)) return;
            if (!DA.GetData(2, ref deg)) return;
            if (deg < 1.0)
            {
                AddRuntimeMessage(GH_RuntimeMessageLevel.Error, "Degree must be greater than 1");
                return;
            }
            if ((Math.Sqrt(deg) - (Math.Floor(Math.Sqrt(deg)))) == 0.0)
            {
                AddRuntimeMessage(GH_RuntimeMessageLevel.Error, "Degree must not be a perfect square");
                return;
            }
            double c = (Math.Sqrt(deg) + 1) / 2;
            Vector3d trans = Vector3d.Unset;
            List<Plane> pointsOut = new List<Plane>();
            List<double> radiiOut = new List<double>();
            for (var i = 0; i < iterations; i++)
            {
                double r = Math.Pow((double)i, c) / (iterations / 2);
                double angle = 2 * Math.PI * c * i;
                double x = (r * Math.Sin(angle) + (iterations));
                double y = (r * Math.Cos(angle) + (iterations));
                Point3d p = new Point3d(x, y, 0.0);
                if (i == 0)
                    trans = (Point3d.Origin - p);
                p.Transform(new Grasshopper.Kernel.Types.Transforms.Translation(trans).ToMatrix());
                Point3d p1 = plane.PointAt(p.X, p.Y, p.Z);
                Plane plane1 = new Plane(p1, plane.XAxis, plane.YAxis);
                pointsOut.Add(plane1);
                double radius = (double)i / (double)iterations;
                radiiOut.Add(radius);
            }
            DA.SetDataList(0, pointsOut);
            DA.SetDataList(1, radiiOut);
        }

        /// <summary>
        /// Provides an Icon for the component.
        /// </summary>
        protected override System.Drawing.Bitmap Icon
        {
            get
            {
                //You can add image files to your project resources and access them like this:
                // return Resources.IconForThisComponent;
                return null;
            }
        }

        /// <summary>
        /// Gets the unique ID for this component. Do not change this ID after release.
        /// </summary>
        public override Guid ComponentGuid
        {
            get { return new Guid("d20b9365-83ed-4636-91ef-da2deaa6e7cd"); }
        }
    }
}